### PR TITLE
[terraform] upgrade Terraform to 0.12.9

### PR DESCRIPTION
**Pull request by scoders**
to: @ryandeivert
cc: @airbnb/streamalert-maintainers
resolves: #947 

## Background
Update streamalert's terraform modules to conform to the latest terraform specifications

## Changes
- Lots of syntax changes to terraform modules introduced by the 0.12 major release 
- Upgraded unit tests to expect explicit relative paths
- Fixed a bug in `streamalert_cli/test/handler.py` while getting non-existent dict key

## Testing
All of the following tests pass:
`tests/scripts/unit_tests.sh`
```
	Ran 1106 tests in 11.219s
	OK
```
`python manage.py test rules`
`python manage.py test classifier`
`python manage.py test live`
```
	Summary:
	Total Tests: 103
	Pass: 103
	Fail: 0
```

### CODE REVIEW

File: streamalert_cli/test/handler.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def __init__(self, options, config):
             'AWS_ACCOUNT_ID': self._config['global']['account']['aws_account_id'],
             'ALERTS_TABLE': '{}_streamalert_alerts'.format(prefix),
         }
-        if options.stats:
+        if 'stats' in options:
```
Comment (ryandeivert) [inline streamalert_cli/test/handler.py]: why this change? `options` is an argparse Namespace object.. not a dict

File: terraform/modules/tf_app_iam/variables.tf [hunk 2]
```diff
@@ -1,9 +1,15 @@
-variable "account_id" {}
+variable "account_id" {
```
Comment (ryandeivert) [inline terraform/modules/tf_app_iam/variables.tf]: did `terraform fmt` do this?

Comment (ryandeivert) [reply terraform/modules/tf_app_iam/variables.tf]: if it did - can we add some `type`s to these

File: terraform/modules/tf_athena/main.tf [hunk 3]
```diff
@@ -160,14 +160,15 @@ resource "aws_cloudwatch_log_group" "athena" {
...
   metric_transformation {
-    name      = "${element(split(",", var.athena_metric_filters[count.index]), 0)}"
-    namespace = "${var.namespace}"
-    value     = "${element(split(",", var.athena_metric_filters[count.index]), 2)}"
+    name      = element(split(",", var.athena_metric_filters[count.index]), 0)
+    namespace = var.namespace
+    value     = element(split(",", var.athena_metric_filters[count.index]), 2)
   }
 }
+
```
Comment (ryandeivert) [inline terraform/modules/tf_athena/main.tf]: a bit of a nit - but can you update your editor to not add extraneous new lines at the bottom of `.tf` files? I don't want us to have to constantly dismiss changes in the future because our IDEs are trying to _remove_ these lines

File: streamalert_cli/test/handler.py (hunk 1 above)
Comment (scoders) [reply streamalert_cli/test/handler.py]: Otherwise I get an error `AttributeError: 'Namespace' object has no attribute 'stats'`
while running  `python manage.py test live` or  `python manage.py test classifier`

File: terraform/modules/tf_kinesis_events/variables.tf [hunk 4]
```diff
@@ -2,12 +2,18 @@ variable "batch_size" {
   default = 100
 }
 
-variable "kinesis_stream_arn" {}
+variable "kinesis_stream_arn" {
```
Comment (ryandeivert) [inline terraform/modules/tf_kinesis_events/variables.tf]: types here too please :)

File: terraform/modules/tf_lookup_tables_s3/main.tf [hunk 5]
```diff
@@ -1,28 +1,45 @@
...
+    actions = ["s3:List*"]
+    # TF-UPGRADE-TODO: In Terraform v0.10 and earlier, it was sometimes necessary to
+    # force an interpolation expression to be interpreted as a list by wrapping it
+    # in an extra set of list brackets. That form was supported for compatibility in
+    # v0.11, but is no longer supported in Terraform v0.12.
+    #
+    # If the expression in the following list itself returns a list, remove the
+    # brackets to avoid interpretation as a list of lists. If the expression
+    # returns a single list item then leave it as-is and remove this TODO comment.
+    resources = [local.s3_bucket_arns]
```
Comment (ryandeivert) [inline terraform/modules/tf_lookup_tables_s3/main.tf]: I think we can just use `local.s3_bucket_arns` here now (removing the `[]` around it)? same as below

File: terraform/modules/tf_cloudtrail/main.tf [hunk 6]
```diff
@@ -1,17 +1,17 @@
 locals {
-  apply_filter_string    = "{ $$.awsRegion != \"${var.region}\" }"
+  apply_filter_string    = "{ $.awsRegion != \"${var.region}\" }"
```
Comment (ryandeivert) [inline terraform/modules/tf_cloudtrail/main.tf]: I actually thing this might be a bad change here - we probably need the double dollar signs

File: terraform/modules/tf_cloudtrail/main.tf [hunk 7]
```diff
@@ -304,9 +304,10 @@ data "aws_iam_policy_document" "cloudtrail_bucket" {
       "s3:PutObject",
     ]
 
-    resources = [
-      "${formatlist("arn:aws:s3:::${local.cloudtrail_bucket_name}/AWSLogs/%s/*", var.account_ids)}",
-    ]
+    resources = formatlist(
+      "arn:aws:s3:::${local.cloudtrail_bucket_name}/AWSLogs/%s/*",
```
[the remark this answers is not in the document]
Comment (scoders) [reply terraform/modules/tf_cloudtrail/main.tf]: done by the autoupgrader

File: terraform/modules/tf_cloudtrail/main.tf (hunk 6 above)
Comment (scoders) [reply terraform/modules/tf_cloudtrail/main.tf]: OK bringing it back

**Comment by scoders-tob** [comment]
Done!

**Comment by ryandeivert** [comment]
merging this into a temporary branch that I will later PR into `release-3-0-0`
